File: src/Mapper/ClientParameterMapper.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Text;
using AutoRest.Core;
using AutoRest.Core.Utilities;
using AutoRest.Extensions;
using AutoRest.Core.Model;
using Newtonsoft.Json;
using AutoRest.Core.Utilities.Collections;
using Newtonsoft.Json.Linq;
using AutoRest.Java.Model;

namespace AutoRest.Java
{
    public class ClientParameterMapper : IMapper<ParameterJv, ClientMethodParameter>
    {
        private ClientParameterMapper()
        {
        }

        private static ClientParameterMapper _instance = new ClientParameterMapper();
        public static ClientParameterMapper Instance => _instance;

        public ClientMethodParameter Map(ParameterJv parameter)
        {
            var settings = JavaSettings.Instance;
            
            IType wireType = Mappers.TypeMapper.Map((IModelTypeJv)parameter.ModelType);
            if (parameter.IsNullable())
            {
                wireType = wireType.AsNullable();
            }
            IType parameterType = wireType.ClientType;

            string parameterDescription = parameter.Documentation;
            if (string.IsNullOrEmpty(parameterDescription))
            {
                parameterDescription = $"the {wireType.ClientType} value";
            }

            IEnumerable<ClassType> parameterAnnotations = settings.NonNullAnnotations && parameter.IsRequired ?
                new[] { ClassType.NonNull } : Enumerable.Empty<ClassType>();

            return new ClientMethodParameter(
                description: parameterDescription,
                isFinal: false,
                wireType: wireType,
                name: parameter.Name,
                isRequired: parameter.IsRequired,
                isConstant: parameter.IsConstant,
                fromClient: parameter.IsClientProperty,
                defaultValue: parameter.DefaultValue,
                annotations: parameterAnnotations);
        }
    }
}

File: javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.autorest.mapper;

import com.azure.autorest.extension.base.model.codemodel.ConstantSchema;
import com.azure.autorest.extension.base.model.codemodel.Parameter;
import com.azure.autorest.extension.base.plugin.JavaSettings;
import com.azure.autorest.model.clientmodel.ClientMethodParameter;
import com.azure.autorest.model.clientmodel.IType;
import com.azure.autorest.model.clientmodel.Versioning;
import com.azure.autorest.util.CodeNamer;
import com.azure.autorest.util.MethodUtil;
import com.azure.autorest.util.SchemaUtil;

import java.util.ArrayList;

/**
 * A mapper that maps an {@link Parameter} to a {@link ClientMethodParameter}.
 */
public class ClientParameterMapper implements IMapper<Parameter, ClientMethodParameter> {
    private static final ClientParameterMapper INSTANCE = new ClientParameterMapper();

    private ClientParameterMapper() {
    }

    /**
     * Gets the global {@link ClientParameterMapper} instance.
     *
     * @return The global {@link ClientParameterMapper} instance.
     */
    public static ClientParameterMapper getInstance() {
        return INSTANCE;
    }

    @Override
    public ClientMethodParameter map(Parameter parameter) {
        return map(parameter, JavaSettings.getInstance().isDataPlaneClient());
    }

    /**
     * Maps an {@link Parameter} to a {@link ClientMethodParameter}.
     *
     * @param parameter The {@link Parameter} being mapped.
     * @param isProtocolMethod Whether the parameter is being used in a protocol method.
     * @return The {@link ClientMethodParameter}.
     */
    public ClientMethodParameter map(Parameter parameter, boolean isProtocolMethod) {
        String name = parameter.getOriginalParameter() != null && parameter.getLanguage().getJava().getName().equals(parameter.getOriginalParameter().getLanguage().getJava().getName())
                ? CodeNamer.toCamelCase(parameter.getOriginalParameter().getSchema().getLanguage().getJava().getName()) + CodeNamer.toPascalCase(parameter.getLanguage().getJava().getName())
                : parameter.getLanguage().getJava().getName();
        name = CodeNamer.getEscapedReservedClientMethodParameterName(name);

        JavaSettings settings = JavaSettings.getInstance();
        ClientMethodParameter.Builder builder = new ClientMethodParameter.Builder()
                .name(name)
                .required(parameter.isRequired())
                .fromClient(parameter.getImplementation() == Parameter.ImplementationLocation.CLIENT);
        if (parameter.getProtocol() != null && parameter.getProtocol().getHttp() != null) {
            builder.requestParameterLocation(parameter.getProtocol().getHttp().getIn());
        }

        IType wireType = Mappers.getSchemaMapper().map(parameter.getSchema());
        if (parameter.isNullable() || !parameter.isRequired()) {
            wireType = wireType.asNullable();
        }
        builder.rawType(wireType);

        if (isProtocolMethod) {
            wireType = SchemaUtil.removeModelFromParameter(parameter.getProtocol().getHttp().getIn(), wireType);
        }

        builder.wireType(wireType)
            .annotations(new ArrayList<>());

        boolean isConstant = false;
        String defaultValue = null;
        if (parameter.getSchema() instanceof ConstantSchema) {
            isConstant = true;
            Object objValue = ((ConstantSchema) parameter.getSchema()).getValue().getValue();
            defaultValue = objValue == null ? null : String.valueOf(objValue);
        }
        builder.constant(isConstant).defaultValue(defaultValue);

        builder.description(MethodUtil.getMethodParameterDescription(parameter, name, isProtocolMethod));

        if (parameter.getExtensions() != null) {
            if (parameter.getExtensions().getXmsVersioningAdded() != null) {
                builder.versioning(new Versioning.Builder()
                        .added(parameter.getExtensions().getXmsVersioningAdded())
                        .build());
            }
        }

        return builder.build();
    }
}
